File: dotnet/source/amp.messaging/MessageSender.cs
using System;
using System.Collections.Generic;
using cmf.bus;
using Common.Logging;

namespace amp.messaging
{
    public class MessageSender
    {
        private static readonly ILog Log = LogManager.GetLogger(typeof(MessageSender));

        private readonly IEnvelopeSender _envelopeSender;
        private readonly IMessageProcessor _messageProcessor;
        

        public MessageSender(IEnvelopeSender envelopeSender)
        {
            _envelopeSender = envelopeSender;
        }

        public MessageSender(IEnvelopeSender envelopeSender, IMessageProcessor messageProcessor)
            : this(envelopeSender)
        {
            _messageProcessor = messageProcessor;
        }

        public MessageSender(IEnvelopeSender envelopeSender, List<IMessageProcessor> processingChain)
            : this (envelopeSender, new MessageProcessorChain(processingChain))
        {
        }

        public void Send(object message)
        {
            this.Send(message, null);
        }

        public void Send(object message, IDictionary<string, string> headers)
        {
            // create an envelope for the message
            Envelope newEnvelope = new Envelope();
            newEnvelope.Headers = headers ?? new Dictionary<string, string>();

            // create a message context for message processing
            MessageContext ctx = new MessageContext(
                MessageContext.Directions.Out, newEnvelope, message);

            // process the message
            this.ProcessMessage(ctx, () =>
            {
                try
                {
                    _envelopeSender.Send(ctx.Envelope);
                }
                catch (Exception ex)
                {
                    string msg = "Failed to send an envelope.";
                    Log.Error(msg, ex);
                    throw new MessageException(msg, ex);
                }
            });
        }


        public void ProcessMessage(
            MessageContext context,
            Action onComplete)
        {
            _messageProcessor.ProcessMessage(context, onComplete);
        }

        public void Dispose()
        {
            _messageProcessor.Dispose();
        }
    }
}

File: java/amp-messaging/src/main/java/amp/messaging/MessageSender.java
package amp.messaging;


import java.util.List;
import java.util.Map;

import cmf.bus.Envelope;
import cmf.bus.IEnvelopeSender;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


/**
 * Created with IntelliJ IDEA.
 * User: jar349
 * Date: 5/1/13
 */
public class MessageSender {

    static final Logger LOG = LoggerFactory.getLogger(MessageSender.class);

    private IEnvelopeSender _envelopeSender;
    private IMessageProcessor _messageProcessor ;

    public IMessageProcessor getMessageProcessor(){
    	return _messageProcessor;
    }

    public MessageSender(IEnvelopeSender envelopeSender) {
        _envelopeSender = envelopeSender;
    }

    public MessageSender(IEnvelopeSender envelopeSender, IMessageProcessor messageProcessor) {
        this(envelopeSender);
        _messageProcessor = messageProcessor;
    }

    public MessageSender(IEnvelopeSender envelopeSender, List<IMessageProcessor> processorChain) {
        this(envelopeSender, new MessageProcessorChain(processorChain));
    }


    public void send(Object message) throws MessageException {

        this.send(message, null);
    }

    public void send(Object message, Map<String, String> headers) throws MessageException {

        if (null == message) { throw new IllegalArgumentException("Cannot send a null message."); }
        LOG.debug("Enter send");

        final Envelope envelope = new Envelope();
        final MessageContext context = new MessageContext(MessageContext.Directions.Out, envelope, message);

        // headers may be null or empty
        if (null != headers) { envelope.setHeaders(headers); }

        try {

            _messageProcessor.processMessage(context, new IContinuationCallback() {

                @Override
                public void continueProcessing() throws MessageException {

                    try {
                        _envelopeSender.send(context.getEnvelope());
                    }
                    catch (Exception ex) {
                        String error = "Failed to send envelope containing message.";
                        LOG.error(error, ex);
                        throw new MessageException(error, ex);
                    }
                }
            });
        }
        catch(Exception ex) {
            String warning = "Caught an exception while processing message.";
            LOG.warn(warning, ex);
            throw new MessageException(warning, ex);
        }

        LOG.debug("Leave send");
    }

    public void dispose() {
    }
}
